### ISSUE TITLE
Unnecessary overhead in placement new [ ]

### ISSUE THREAD

**Issue by HPaulS** [cplusplus/draft#2264]
With regard to the above, I can't think of a single good reason why thevsrandard continues to say this (emphasis  mine):

This overhead may be applied in all array new-expressions, __including those referencing the library function operator new[](std​::​size_­t, void*) and other placement allocation functions.__

Link: http://eel.is/c++draft/expr.new#15

Can this be changed?

**Comment by jwakely** [comment]
This doesn't sound like an editorial change.

**Comment by HPaulS** [comment]
Oh sorry, is that not what we're meant to be posting here?  It was intended as a request to amend the standard but if such a request is inappropriate here then I will take it elsewhere.

**Comment by jwakely** [comment]
See https://github.com/cplusplus/draft/wiki/How-to-tell-if-an-issue-is-editorial

**Comment by tkoeppe** [comment]
Please also browse the existing issues lists, e.g. http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_closed.html#476.

It sounds like you need to send a paper to EWG if you want to make progress here.

**Comment by HPaulS** [comment]
OK, thanks for digging that up, useful to know.  I'm not sure I have the energy to write a _paper_ (and I'm not sure I'd call this 'evolution' anyway) so I might just leave it there and see if any of the standards guys pick up on this.  That doesn't seem _un_ reasonable to me, but we'll see.

**Comment by HPaulS** [comment]
Oh my, what a mess.  In the link _you_ gave (for which much thanks), I read (emphasis mine):

> EWG thinks there's no design defect here, and suggests that a placement array new should __not__ have such implementation-defined overhead. EWG suggests that CWG should clarify that __placement array new T[n] needs n*sizeof(T) bytes of memory__, and clarify that an __operator delete must only be called if the corresponding operator new has been called__.

That being a non-placement new, of course.  So they agree with me, but they passed the buck to the CWG.

But in the link linked to from _there_ (core issue 476, which has already been pointed out to me) we see, along with a lot of other irrelevant verbiage:

> It has been suggested that this issue is more appropriate for EWG and should thus be closed with "extension" status.

Which, frankly, is a load of cr@p.  And that's where the trail goes cold.  So that worked (!)

So where does that leave us?  Is there a way I can reopen this because I'm not prepared to put any further effort into it.  If that has no effect, fine, but at least I tried.  And again, to be frank, I thought you were premature and a bit cavalier in closing it in the first place, no matter what your status.  Thank you.

**Comment by tkoeppe** [comment]
@HPaulS: No matter the relevance of the problem, please bear in mind that this repository and this issue tracker are strictly for *editorial* work. Everything else here is distracting noise for the already-busy editors. There are better ways to report and discuss technical problems with the standard.

**Comment by HPaulS** [comment]
@tkoeppe I don't believe this is noise.  Like I said, I've done all I'm prepared to do and I'm done with it.

**Comment by jwakely** [comment]
This is neither the place to report technical issues nor to discuss them. It was reported to CWG, who sent it to EWG, who said it's not their problem, and CWG never re-opened it. If you want to ask them to re-open it then this is the wrong place to do so.

**Comment by HPaulS** [comment]
@jwakely Cool, thanks. I've done that (and referenced this thread, which I believe will continue to stick around).   I didn't realise it was so easy, I owe you one.  Have a great day.

**Comment by HPaulS** [comment]
Update: Well, of course, this needs to be reported to the __EWG__, not the CWG, silly me.  Anyway, it seems that Thomas has been doing his best to get this fixed already (I guess he just forgot).  I include below, just for completeness, a transcript of my email to the owner of the EWG closed issues list which references Thomas's (excellent) defect report.  Cheers all.

-----

Hi there,

As you are the owner of the EWG 'closed issues' list, I am writing in reference to this:

https://cplusplus.github.io/EWG/ewg-closed.html#68

Is there anything happening with this please?  This is an important issue and it's just being swept under the carpet for some reason.

It was also reported again a bit later on, see:

https://cplusplus.github.io/EWG/ewg-closed.html#90

and:

http://www.open-std.org/jtc1/sc22/wg21/docs/cwg_closed.html#476

So it got bounced around between the CWG and the EWG and then, I guess, it just got forgotten about.  Nice.

So _someone_ needs to stop saying ridiculous things like this:

> In Chicago 2013, EWG deemed this NAD. Remarks from the discussion: "The problem is in trying to use array new to put an array into pre-existing storage. We don't need to use array new for that; just construct them." and "You can find what padding was used after the fact, because new returns a pointer to the start of the first element of the array. But that can happen only after UB. "

and get on and do the job properly.  You can see there that Thomas is doing his best but noone appears to be listening to him.

My take on it here:

https://stackoverflow.com/a/51362295/5743288

Many thanks and best wishes - Paul Sanders.

**Comment by FrankHB** [comment]
You may better discuss it at [std-discussion](https://groups.google.com/a/isocpp.org/forum/?fromgroups#!forum/std-discussion).

**Comment by tkoeppe** [comment]
@HPaulS: I hadn't forgotten, but I also wasn't able to find the issue for some reason. Thanks for digging that one up. That was my first involvement with the standardization process as far as I can remember; I thought that feature can't possibly make sense, why don't we just fix that...

**Comment by HPaulS** [comment]
OK, Thomas.  And sorry Frank, I do know about that list but I'm not prepared to get involved in any sort of extended discussion about this as I believe I already said.  Thomas already poured his heart and soul into this back in the day and it got him nowhere.  And the guys here are trying to keep the noise down.

**Comment by tkoeppe** [comment]
Well, as an Editor I'm also one of the guys who want to keep the noise down :-) But I could try and ask around if we still agree with the assessment from back then. I believe the CWG's position at the time was that while the *standard* placement-array-new doesn't require an overhead, there may be *overloads* that do, and CWG were reluctant to change only one particular part of this feature ad-hoc. They preferred if EWG looked at the entire feature holistically. I'm sympathetic to that sentiment, and I think the best way would be an EWG paper. It wouldn't need to be long, but a paper can always be listened to, whereas issues may not end up progressing if there's no clear agreement on direction.

So, in summary, I think the problem is real, the issues lists aren't sufficient to address this, a paper is needed, and this is not the right place for that work. The std-* mailing lists are a good place to start if you want to get involved, but without someone driving this I don't expect anything will happen, and most people who are already busy with standards work probably have more urgent things they want to get done.

**Comment by jwakely** [comment]
Right, if you think it's important you need to work on it, not just tell other people to work on it because you think it's important. That's not how these things get done.

And the stackoverflow question seems a better place for updates, it really doesn't belong here.

**Comment by HPaulS** [comment]
@All of you.  As I believe I mentioned, I emailed the guy who manages the EWG closed issue list and I (obviously) don't seek to prolong any discussion of the matter here.  I just posted my email here __for reference__, just in case anyone (e.g. you, Thomas) was interested.  What happens is out of my hands now.  I think I said that, too.

And, if you _read_ that email (which I posted, remember, _for reference_), you will see that I linked directly to Thomas's paper, which, IMO, sums up the problem nicely (and certainly far better than I ever could).  One thing's for sure: we certainly don't need another one.  So, I say again, I see no need for _any of us_ to do any more, and most certainly not in this thread.  Que sera sera, I'm losing interest.  Thank you.

OK, one thing:

Don't get bogged down in the detail Thomas.  Take it from me, placement new [ ] doesn't need that overhead.  You know it, and I know it.  The problem is convincing the committee, and you already did your best to do that already.  Let it go brother, let it go.  I already shook the tree with that email, so, we'll see.

And, FWIW:  My 'trying keep the noise down' rider was (obliquely) aimed at Frank, not you.  Perhaps I shouldn't have.

@jwakely Don't be daft!  I have no idea what you're trying to say to me there.

**Comment by HPaulS** [comment]
> I thought that feature can't possibly make sense, why don't we just fix that...

Sorry Thomas, what was I thinking?  Yes, my thoughts exactly.

Sorry about that - Paul.

**Comment by jwakely** [comment]
I have no idea what you're on about. I suggested you ask CWG to reopen the issue, then realised that it would be helpful to let you know how to do that, pointing you to the email address. I don't understand your problem with that, but I also don't care. I'm muting this thread now, so I don't see any more of your twaddle.

**Comment by HPaulS** [comment]
No worries.  Absolute power corrupts, absolutely.

**Comment by dhouck** [comment]
No, in this case it seems to be miscommunication exacerbated by both of you
being frustrated with the other. I think jwakely always thought the CWG was
the right place for this, not EWG.

In any case, you say you don't want to move to std-discuss because you
don't want to be involved with it anymore, but you then keep posting here.
If you don't want to be involved any more, don't be; if you do, the right
place for anything from status report up is std-discuss. Even though
everybody here seems to agree this is a problem to some extent, this is
only a slightly better place to discuss it than it is to discuss undefined
behavior in Rust.

There has also been no paper about this, by Thomas or anybody else. A paper
is a formal document with a p-number or an n-number, like [p0593](
http://wg21.link/p0593) (which, incidentally, solves or provides a
workaround for this depending on use case). I agree with Thomas that
creating a paper is the next step for this issue.

My recommendation is that, if you have any other updates, you make a
summary post of this issue on std-discuss instead of here and just post a
link to that here (and in the comments to my SO answer about realloc or the
other question about this specifically). That is not a promise to put any
further work into solving this

**Comment by HPaulS** [comment]
This is just bullshit.  And I thought this thread had been muted.

**Comment by zygoloid** [comment]
@HPaulS I appreciate that this may be frustrating you. However, your choice of language and hostility towards others in this thread is unacceptable. Please conduct yourself with professionalism, or I will block you from commenting here further.

**Comment by HPaulS** [comment]
OK, maybe I was a bit out of line there, but __Thomas has done all the legwork__ and nobody seems to be listening.  Just for those who don't remember the link, it's here:

https://cplusplus.github.io/EWG/ewg-closed.html#68

And that's all I have to say, because the actions of Jonathan and the owners of the two closed issue lists say clearly to me that this is all just a cesspit and I am sick to the back teeth with it.  You can do what the hell you like.

**Comment by villevoutilainen** [comment]
Funny, the owners of the two closed issue lists agreed to reopen the matter.

**Comment by HPaulS** [comment]
@ville Cool, thanks for the heads-up.  Nice one guys, hats off.

**Comment by jwakely** [comment]
Your latest rants have been brought to my attention, and I'd continue to ignore you but for ...

> And I thought this thread had been muted.

I muted it **for me** so I don't get email updates and I don't have to read your displays of personality defects and attitude problems. I didn't lock the thread or prevent you or anybody else from participating, although I see you're getting close to achieving that without my help.

**Comment by HPaulS** [comment]
@ville Seems so, apologies.  Anyway, what matters is getting the right result, no?

**Comment by HPaulS** [comment]
@jwakely  Is it just me that you seem to like to vent your spleen on, or is it everyone?  Whatever it is it's not very nice, so please just cut it out.  Thx.

**Comment by jfbastien** [comment]
I think it's just you. 👍
